### PR TITLE
Checking .proto, price/endpoint in service json and Agent Contract

### PR DESCRIPTION
**Pull request by arturgontijo**
- Remove some newlines (code structure).
- Exception at all `transact()` in `publish()`.
- All found .proto are compiled to check if they are valid before publish/update (Related #48).
- `snet` checks if `price` field is a valid number.
- `snet` sends a GET request to endpoint before publish/update (Related #49).
- Check the current state of the Agent Contract [L990-L1009] (Related #52).

### CODE REVIEW

File: snet_cli/commands.py [hunk 1]
```diff
@@ -813,6 +812,102 @@ def _getserviceregistrationbyname(self, service_json=None):
...
+                entry_path = Path.cwd().joinpath(spec_path).resolve()
+            else:
+                entry_path = spec_path.resolve()
+            if not os.path.isdir(entry_path):
+                self._printerr("Service spec path must resolve to a valid directory: {}".format(spec_path))
+                is_ok = False
+            import_paths = walk_imports(entry_path)
+
+            if import_paths:
+                codegen_tmp_dir = Path("~").expanduser().joinpath(".snet").joinpath("py-codegen").joinpath(
```
Comment (vsbogd) [inline snet_cli/commands.py]: I would use [tempfile.mkdtemp](https://docs.python.org/2/library/tempfile.html#tempfile.mkdtemp) instead

File: snet_cli/commands.py [hunk 2]
```diff
@@ -813,6 +812,102 @@ def _getserviceregistrationbyname(self, service_json=None):
...
+        metadata_ipfs_uri = uri_reference(metadata_ipfs_path).copy_with(scheme='ipfs').unsplit()
+
+        # Checking price
+        if not type(service_json["price"]) == int:
+            self._printerr("Invalid price format: {} [{}]".format(service_json["price"], type(service_json["price"])))
+            is_ok = False
+
+        # Checking endpoint
+        try:
+            request = requests.get(str(service_json["endpoint"]), timeout=5)
```
Comment (vsbogd) [inline snet_cli/commands.py]: Do not think we should actually check that endpoint is live as it prevents us from publishing service which is temporary down. Is it enough to check that endpoint has correct URI format?

File: snet_cli/commands.py (hunk 1 above)
Comment (vsbogd) [inline snet_cli/commands.py]: BTW it seems this code duplicates https://github.com/singnet/snet-cli/blob/fdbaea78e566c6d470996a3243fe288f9532a94d/snet_cli/commands.py#L518

File: snet_cli/commands.py [hunk 3]
```diff
@@ -813,6 +812,102 @@ def _getserviceregistrationbyname(self, service_json=None):
...
+                # stdout back to console
+                sys.stdout = sys.__stdout__
+                if os.path.exists(codegen_tmp_dir):
+                    rmtree(codegen_tmp_dir)
+
+        # Create service_spec/ tar and upload it to IPFS
+        ipfs_endpoint = urlparse(self.config["ipfs"]["default_ipfs_endpoint"])
+        ipfs_scheme = ipfs_endpoint.scheme if ipfs_endpoint.scheme else "http"
+        ipfs_port = ipfs_endpoint.port if ipfs_endpoint.port else 5001
+        ipfs_client = ipfsapi.connect(urljoin(ipfs_scheme, ipfs_endpoint.hostname), ipfs_port)
```
Comment (vsbogd) [inline snet_cli/commands.py]: This IPFS client code is copied three times as far as I see. Could you please wrap it to separate function and reuse.

File: snet_cli/commands.py (hunk 2 above)
Comment (arturgontijo) [reply snet_cli/commands.py]: Even if the endpoint is unreachable the user can type "y" and proceed with the publishing/update.
If we check it with something like `urlparse` (is this what you're suggesting?) some mistyping can pass unnoticed, like: http://singulritynet.io:7000.

File: snet_cli/commands.py (hunk 1 above)
Comment (arturgontijo) [reply snet_cli/commands.py]: I'll create a function to use this `protoc` part of code.

File: snet_cli/commands.py (hunk 3 above)
Comment (arturgontijo) [reply snet_cli/commands.py]: Sure!

File: snet_cli/commands.py (hunk 2 above)
Comment (vsbogd) [reply snet_cli/commands.py]: Ah ok didn't catch it at first time. If we can proceed even if service is down that is ok I think. Thanks!

**Comment by vsbogd** [comment]
Looks good, thanks!

**Pull request merged**